client/src/main/java/io/pravega/client/admin/impl/ReaderGroupManagerImpl.java

```diff
@@ -75,28 +70,29 @@ public ReaderGroupManagerImpl(String scope, Controller controller, AbstractClien
-        getAndHandleExceptions(controller.createStream(scope, streamName, StreamConfiguration.builder()
-                                                                          .scalingPolicy(config.getScalingPolicy())
-                                                                          .build()),
-                               RuntimeException::new);
-        return new StreamImpl(scope, streamName);
-    }
-
     @Override
     public void createReaderGroup(String groupName, ReaderGroupConfig config) {
         log.info("Creating reader group: {} for streams: {} with configuration: {}", groupName,
                 Arrays.toString(config.getStartingStreamCuts().keySet().toArray()), config);
         NameUtils.validateReaderGroupName(groupName);
-        createStreamHelper(getStreamForReaderGroup(groupName), StreamConfiguration.builder()
-                                                                                  .scalingPolicy(ScalingPolicy.fixed(1))
-                                                                                  .build());
-        ReaderGroupConfig ctrlConfig = getThrowingException(controller.createReaderGroup(scope, groupName, config));
+        if (config.getReaderGroupId() == ReaderGroupConfig.DEFAULT_UUID) {
+            // make sure we never attempt to create a ReaderGroup with default ReadrGroupId which is 0-0-0
+            config = config.toBuilder().readerGroupId(UUID.randomUUID()).generation(0L).build();
+        }
+        ReaderGroupConfig controllerConfig = getThrowingException(controller.createReaderGroup(scope, groupName, config));
         @Cleanup
         StateSynchronizer<ReaderGroupState> synchronizer = clientFactory.createStateSynchronizer(NameUtils.getStreamForReaderGroup(groupName),
                                               new ReaderGroupStateUpdatesSerializer(), new ReaderGroupStateInitSerializer(), SynchronizerConfig.builder().build());
-        Map<SegmentWithRange, Long> segments = ReaderGroupImpl.getSegmentsForStreams(controller, ctrlConfig);
-        synchronizer.initialize(new ReaderGroupState.ReaderGroupStateInit(ctrlConfig, segments, getEndSegmentsForStreams(ctrlConfig), false));
+        Map<SegmentWithRange, Long> segments = ReaderGroupImpl.getSegmentsForStreams(controller, controllerConfig);
+        synchronizer.initialize(new ReaderGroupState.ReaderGroupStateInit(controllerConfig, segments, getEndSegmentsForStreams(controllerConfig), false));
+        ReaderGroupConfig syncConfig = synchronizer.getState().getConfig();
+        if (ReaderGroupConfig.DEFAULT_UUID.equals(syncConfig.getReaderGroupId())
+        && ReaderGroupConfig.DEFAULT_GENERATION == syncConfig.getGeneration()) {
+            // migrate RG from version < 0.9 to 0.9+
+            synchronizer.updateState((s, updates) -> {
+              updates.add(new ReaderGroupState.ReaderGroupStateInit(controllerConfig, segments, getEndSegmentsForStreams(controllerConfig), false));
```

Comment:
Initialize does not take effect if StateSynchronizer is already present.
When initialize does take effect, syncConfig.getReaderGroupId() cannot be DEFAULT_ID.

Reply:
so what i dont completely understand is line 87 is also doing synchronizer.initialize with RGstate.init.
so what are we doing post line 88. we have already recorded the config in controller and attempted to write into state synchronizer if its not present. 
then we get the object in state synchronizer. there will definitely be something, either an older config or the one we just wrote.. 
 
if the state synchronizer already had older config which when deserialized into new object will have default id we enter the if block.. 
and we call update with RGInit inside the update block.. will this update the RGConfig in state sync or not?

Reply:
`synchronizer.initialize()`  passes the `oldestRevision` to `client.writeConditionally()` due to which it will never actually perform the write except the very first time.
where as `synchronizer.updateState()` passes `revision` inside the ReaderGroupState object which should be close to the latest, if not the latest.... so it will eventually update the Stream.
As discussed, I will remove the if block from `createReaderGroup()` and we can rely on `resetReaderGroup` to be invoked for RG Migration.

Reply:
Removed the "if" check for DEFAULT_UUID from createReaderGroup, as discussed.